### ISSUE TITLE
Influence of Batch normalization

### ISSUE THREAD

**Issue by NinaJacobsen** [Kamnitsask/deepmedic#118]
Hi Kamnitsas,
I do realize my last thread with a similar topic was closed recently, but I have a related question, which might help mutual users as well.

I did an experiment training a model (with more or less the original deepMedic parameters) changing only the setting for batch normalization, having it "switched on" and "switched off" (set = 0). From theory, my expectations to the results would have been, that the training progress would have become slower (the progress should converge later) as I turned off the BN - however, the training progresses for the two networks looks exactly the same, suggesting BN had no influence. My data were normalized to zero mean unit variance using white stripe method.

Have you tried this experiment? What result would you expect?

Thanks,
Nina.

**Comment by ChaitanyaBaweja** [comment]
This has been addressed to a certain extent in his paper [Efficient multi-scale 3D CNN with fully connected CRF for accurate brain lesion segmentation](https://www.sciencedirect.com/science/article/pii/S1361841516301839). The author writes : 

>  For countering it, we adopt the recently proposed Batch Normalisation (BN) technique to all hidden layers (Ioffe and Szegedy, 2015), which allows normalization of the FM activations at every optimization step in order to better preserve the signal.

By it, the author means the general problems associated with training a deep network like vanishing gradients etc. Batch normalization also adds a weak regularization which is usually helpful. 

Why your training time hasn't improved could be due to a lot of reasons. In my case, the bottleneck is the sampling process which won't be really affected by batch normalization.

**Comment by Kamnitsask** [comment]
Hello, and sorry for the reply. I ve been on a trip away from office lately and couldnt follow everything.

On BN:
In different experimental settings, effect of BN "may" be seen in two points: It may accelerate training and it may increase final accuracy.

But, I would expect that whether these effects will be seen and how strongly depends on the task & experimental settings (also suggested above by Chaitanya). 

For example, if the settings are relatively well configured and maybe the task relatively easy to learn, then BN may not cause significant effect. If a task is difficult to learn, then its effect may become more obvious. For example, in my experiments with TBI (MedIA journal), BN clearly helped as shown. On multi-organ segmentation where there is massive class imbalance, BN also helped to train better. In some side-experiments in a more later version of DM on brain segmentation, where learning seems easier, BN didnt change things much.

Other factors that influence convergence and their effects are interleaved when observing convergence curves can be: learning rate schedule, optimizer (sgd/adam/rms etc), regularizers (dropout, L1, L2 etc), augm, etc. 
 
I hope the above help. And sorry again for the late reply, doing my best.

**Comment by Kamnitsask** [comment]
I had a look at a couple of tasks (brain segmentation and WML segm) with / without BN. Convergence speed looks more or less the same both for train & val for both tasks. The only difference seems to be in the overall accuracy (mean acc on samplse and DSC, both on samples and whole volumes). BN improves acc by a bit, especially visible on DSC, I guess due to its regularization effect.

Hope this helps.
K

**Comment by NinaJacobsen** [comment]
Thanks for the answer, helped a lot!

A related question. I recently stumbled upon "adaptive batch normalization" by Li et al, who are claiming their method can help domain adaption. Basically, they suggest to apply BN in the testing phase and use mean and variance from the target domain (training data) to adjust the data of the source domain throughout the network.

Have you tested this method or considered implementing it?

NJ

**Comment by Kamnitsask** [comment]
Hello,

I wont comment too much further on this BN for DA because these issues are not for such discussions.
In short, this method of aligning the 2 first moments (mean/std) of the 2 distributions for DA has been tried before, adaBN is not first. It was reported by various people that such an approach alleviates the problem somewhat, it's not a solution. (just adjusting the means and the std may not be sufficient in difficult tasks).
That's why other, more flexible methods are being focused on (eg GANs, MMD, etc, although difficult to make them work).
I had tried it a year ago, it did something positive, but far from sufficient. It may be useful within a larger DA framework, but not sufficient by itself in problems with difficult domain-shift to my experience. Perhaps in simple problems?

At some point I want to do a repass on BN of DeepMedic, and at that point such methods will probably be implemented. But I keep on failing to make time for that, so dont count on it for your own project/plans. I am very behind with some things I have promised before :(

Kind regards,
Kostas
